### PR TITLE
feat: integration open frame

### PR DESCRIPTION
**Pull request by dudu0506**
# Pull Request

## Related Issue

Closes #0000

## Checklist

-   [ ] I have tested these changes locally.
-   [ ] I have reviewed the code changes.
-   [ ] I have added unit tests if applicable.

### CODE REVIEW

File: constants/zod.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { z } from 'zod';
+
+import { IMAGE_THEME } from '@/constants/theme';
+
+export const IMAGE_QUERY_SCHEMA = z.object({
+    id: z.string(),
+    profileId: z.string().optional(),
+    date: z.string().optional(),
+    theme: z.enum([IMAGE_THEME.Dark, IMAGE_THEME.Light]).optional().default(IMAGE_THEME.Light),
```
Comment (guanbinrui) [inline constants/zod.ts:9]:
```suggestion
    theme: z.nativeEnum(IMAGE_THEME).optional().default(IMAGE_THEME.Light),
```

File: helpers/getPollFrameImage.tsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import { ErrorHolder } from '@/components/ErrorHolder';
+import { PollCard } from '@/components/PollCard';
+import { IMAGE_THEME } from '@/constants/theme';
+
+interface Parameters {
+    poll: Poll | null;
+    theme: IMAGE_THEME;
+    profileId?: string;
+    newVotedIdx?: number;
+}
```
Comment (guanbinrui) [inline helpers/getPollFrameImage.tsx:11]:
```suggestion
interface Parameters {
    poll: Poll;
    theme: IMAGE_THEME;
    profileId?: string;
    newVotedIdx?: number;
}
```

File: helpers/isCreatedByProfileId.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { Poll } from '@/app/types';
+
+// FIXME: Maybe we should refactor this compare logic
+export const isCreatedByProfileId = (poll: Poll, profileId?: string) => {
+    return poll.created_by === profileId;
```
Comment (guanbinrui) [inline helpers/isCreatedByProfileId.ts:5]:
```suggestion
    return !!profileId && poll.created_by === profileId;
```

File: helpers/getPollFrameButtons.tsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+import { Poll } from '@/app/types';
+import { PER_USER_VOTE_LIMIT } from '@/constants';
+import { POLL_STATUS } from '@/constants/enum';
+import { ImageQuery } from '@/constants/zod';
+import { getPollFramePostUrl } from '@/helpers/getPollFramePostUrl';
+import { isCreatedByProfileId } from '@/helpers/isCreatedByProfileId';
+
+interface Parameters {
+    poll: Poll | null;
```
Comment (guanbinrui) [inline helpers/getPollFrameButtons.tsx:11]:
```suggestion
    poll: Poll;
```

File: components/FrameViewer.tsx [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    useEffect(() => {
+        const imageMeta = document.querySelector<HTMLMetaElement>('meta[name="of:image"]');
+        setImg(imageMeta?.getAttribute('content') ?? null);
+    }, []);
+
+    if (IS_PRODUCTION) {
+        return null;
+    }
+
+    return <div>{img ? <img src={img} /> : null}</div>;
```
Comment (guanbinrui) [inline components/FrameViewer.tsx:20]:
```suggestion
    if (!img) return null;

    return <div>{img ? <img src={img} /> : null}</div>;
```

File: components/FrameViewer.tsx [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import { IS_PRODUCTION } from '@/constants';
+
+// NOTICE: Just for development
+export function FrameViewer() {
+    const [img, setImg] = useState<string | null>('');
+
+    useEffect(() => {
+        const imageMeta = document.querySelector<HTMLMetaElement>('meta[name="of:image"]');
+        setImg(imageMeta?.getAttribute('content') ?? null);
+    }, []);
```
Comment (guanbinrui) [inline components/FrameViewer.tsx:14]:
```suggestion
    const img = imageMeta?.getAttribute('content') ?? null;
```

File: app/form.tsx [hunk 7]
```diff
@@ -31,21 +32,9 @@ export function PollCreateForm() {
...
-        votes1: 0,
-        votes2: 0,
-        votes3: 0,
-        votes4: 0,
-        validInDays: MIN_VALID_IN_DAYS,
-    };
-    const saveWithNewPoll = savePoll.bind(null, pollStub);
+    const pollStub = createNullPoll();
+    pollStub.id = uuid();
+    pollStub.created_at = Date.now();
```
Comment (guanbinrui) [inline app/form.tsx:37]: Assign id and created_at in `createNullPoll`.

```suggestion
    const pollStub = createNullPoll();
    pollStub.id = uuid();
    pollStub.created_at = Date.now();
```

File: app/form.tsx [hunk 8]
```diff
@@ -31,21 +32,9 @@ export function PollCreateForm() {
...
-        option3: '',
-        option4: '',
-        votes1: 0,
-        votes2: 0,
-        votes3: 0,
-        votes4: 0,
-        validInDays: MIN_VALID_IN_DAYS,
-    };
-    const saveWithNewPoll = savePoll.bind(null, pollStub);
+    const pollStub = createNullPoll();
```
Comment (guanbinrui) [inline app/form.tsx:35]: Sure to create poll for each render?

File: constants/env.ts [hunk 9]
```diff
@@ -0,0 +1,5 @@
+export const HOST = process.env.HOST;
```
Comment (guanbinrui) [inline constants/env.ts:1]: Use safe env like: https://github.com/DimensionDev/firefly.mask.social/blob/main/src/constants/env.ts

File: helpers/getPollFrameImage.tsx [hunk 10]
```diff
@@ -0,0 +1,18 @@
+import { Poll } from '@/app/types';
+import { ErrorHolder } from '@/components/ErrorHolder';
+import { PollCard } from '@/components/PollCard';
+import { IMAGE_THEME } from '@/constants/theme';
+
+interface Parameters {
+    poll: Poll | null;
+    theme: IMAGE_THEME;
```
Comment (guanbinrui) [inline helpers/getPollFrameImage.tsx:8]: Request for supporting client locale.

```suggestion
    theme: IMAGE_THEME;
    locale: IMAGE_LOCALE;
```

File: helpers/getPollFrameButtons.tsx [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    const postTarget = getPollFramePostUrl(queryData);
+    const votedIdxList =
+        poll?.options.reduce(
+            (acc, opt, index) => {
+                return opt.voted ? [...acc, index + 1] : acc;
+            },
+            newVotedIdx ? [newVotedIdx] : [],
+        ) ?? [];
+    if (
+        !poll ||
```
Comment (guanbinrui) [inline helpers/getPollFrameButtons.tsx:27]:
```suggestion
        !poll ||
```
```suggestion
```

File: constants/index.ts [hunk 12]
```diff
@@ -11,17 +14,12 @@ export const POLL_EXPIRY = 60 * 60 * 24 * 180; // Expire polls after 3 months
...
-    votes3: 0,
-    votes4: 0,
-    created_at: 0,
-    validInDays: MIN_VALID_IN_DAYS,
-};
+export const DEFAULT_FRAME_VERSION = 'vNext';
+
+export const PER_USER_VOTE_LIMIT = 1;
+
+export const COMMON_APP_TITLE = 'Frame Server';
```
Comment (guanbinrui) [inline constants/index.ts:21]:
```suggestion
export const COMMON_APP_TITLE = 'Polls';
```

File: src/services/getPoll.ts [hunk 13]
```diff
@@ -0,0 +1,8 @@
+import { kv } from '@vercel/kv';
+
+import { Poll } from '@/types';
+
+export const getPoll = async (pollId: string, profileId?: string): Promise<Poll | null> => {
+    const poll = await kv.hgetall<Poll>(`poll:${pollId}`);
```
Comment (guanbinrui) [inline src/services/getPoll.ts:6]:
```suggestion
    const poll = await kv.hgetall<Poll>(`poll:${poll.source}_${pollId}`);
```

File: src/middleware.ts
Comment (guanbinrui) [inline src/middleware.ts:1]: Delete empty file.

File: src/app/api/frames/vote/route.tsx [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+import { FRAME_SOURCE, POLL_STATUS } from '@/constants/enum';
+import { IMAGE_QUERY_SCHEMA } from '@/constants/zod';
+import { getPollFrameButtons } from '@/helpers/getPollFrameButtons';
+import { getPollFrameImage } from '@/helpers/getPollFrameImage';
+import { isCreatedByProfileId } from '@/helpers/isCreatedByProfileId';
+import { resolveFrameSource } from '@/helpers/resolveFrameSource';
+import { getPoll } from '@/services/getPoll';
+import { vote } from '@/services/vote';
+
+export const POST = frames(async (ctx) => {
```
Comment (guanbinrui) [inline src/app/api/frames/vote/route.tsx:15]: Need validate ctx?

File: src/app/api/frames/vote/route.tsx [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+    if (!profileId || buttonIndex < 1 || buttonIndex > POLL_OPTIONS_MAX_COUNT) {
+        error(`Got invalid parameters: profileId=${profileId}, buttonIndex=${buttonIndex}`, StatusCodes.BAD_REQUEST);
+    }
+
+    const queryData = IMAGE_QUERY_SCHEMA.parse(ctx.searchParams);
+    const { id: pollId, theme } = queryData;
+    if (!pollId) {
+        error(`Missing PollId in frame state`, StatusCodes.BAD_REQUEST);
+    }
```
Comment (guanbinrui) [inline src/app/api/frames/vote/route.tsx:39]: Duplicate with `IMAGE_QUERY_SCHEMA`.

File: src/app/api/frames/vote/route.tsx [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+
+    if (poll.status !== POLL_STATUS.Active) {
+        error(`The poll is unavailable via pollId = ${pollId}`);
+    }
+
+    const votedLen = poll.options.filter((opt) => opt.voted).length;
+    if (votedLen >= PER_USER_VOTE_LIMIT) {
+        error(`You have voted ${votedLen} times, cannot vote again`);
+    }
```
Comment (guanbinrui) [inline src/app/api/frames/vote/route.tsx:53]: Validate with zod?

File: src/app/page.tsx [hunk 17]
```diff
@@ -0,0 +1,14 @@
+import { PollCreator } from '@/components/PollCreator';
+
+export const metadata = {
+    title: 'Farcaster polls',
+    description: 'Poll example for farcaster',
```
Comment (guanbinrui) [inline src/app/page.tsx:5]:
```suggestion
    title: 'Polls',
    description: 'Create and share your polls with Frame.',
```

File: src/components/FrameViewer.tsx [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+    const frameData = collectFrameDataFromDocument();
+
+    if (IS_PRODUCTION || !frameData) {
+        return null;
+    }
+
+    return (
+        <div className="flex h-screen w-screen flex-col items-center justify-center">
+            <div className="h-[400px] w-[600px]">
+                <img src={frameData.image} />
```
Comment (guanbinrui) [inline src/components/FrameViewer.tsx:17]: Alt message if possible.

File: src/components/PollCard.tsx [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+            <div
+                style={{
+                    display: 'flex',
+                    flexDirection: 'column',
+                    width: '100%',
+                    gap: 12,
+                    padding: 20,
+                }}
+            >
+                {/* <h2 style={{ textAlign: 'center', color: themeConfig.titleColor }}>{title}</h2> */}
```
Comment (guanbinrui) [inline src/components/PollCard.tsx:143]:
```suggestion
```

File: src/helpers/resolveFrameSource.ts [hunk 20]
```diff
@@ -0,0 +1,12 @@
+import { FRAME_SOURCE } from '@/constants/enum';
+
+export const resolveFrameSource = (frameSource: string): FRAME_SOURCE | null => {
```
Comment (guanbinrui) [inline src/helpers/resolveFrameSource.ts:3]: Just force casting.

File: src/constants/env.ts [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+import { NODE_ENV, VERCEL_NEV } from '@/constants/enum';
+
+const InternalEnvSchema = z.object({});
+
+const ExternalEnvSchema = z.object({
+    NEXT_PUBLIC_VERCEL_ENV: z.nativeEnum(VERCEL_NEV).default(VERCEL_NEV.Development),
+    NEXT_PUBLIC_FIREFLY_API_URL: z.string().default('https://api.firefly.land'),
+    HOST: z.string(),
+    HUB_URL: z.string(),
+    HUB_API_KEY: z.string().optional(),
```
Comment (guanbinrui) [inline src/constants/env.ts:12]:
```suggestion
    NEXT_PUBLIC_HOST: z.string(),
    NEXT_PUBLIC_HUB_URL: z.string(),
    NEXT_PUBLIC_HUB_API_KEY: z.string().optional(),
```

File: src/app/api/frames/route.tsx [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+import { createFrameSuccessResponse } from '@/helpers/createFrameSuccessResponse';
+import { createFrameTranslator } from '@/helpers/createFrameTranslator';
+import { getPoll } from '@/services/getPoll';
+
+const handleRequest = frames(async (ctx) => {
+    let queryData: ImageQuery | null = null;
+    try {
+        queryData = IMAGE_QUERY_SCHEMA.parse(ctx.searchParams);
+        const { id, profileId, source, locale } = queryData;
+        const t = createFrameTranslator(locale);
```
Comment (guanbinrui) [inline src/app/api/frames/route.tsx:13]: Move to `locales/index.ts`.

```ts
export const t = createFrameTranslator(locale);
```

File: src/app/api/frames/route.tsx [hunk 23]
```diff
@@ -0,0 +1,30 @@
+import { frames } from '@/config/frames';
+import { IMAGE_QUERY_SCHEMA, ImageQuery } from '@/constants/zod';
+import { createFrameErrorResponse } from '@/helpers/createFrameErrorResponse';
+import { createFrameSuccessResponse } from '@/helpers/createFrameSuccessResponse';
+import { createFrameTranslator } from '@/helpers/createFrameTranslator';
+import { getPoll } from '@/services/getPoll';
+
+const handleRequest = frames(async (ctx) => {
+    let queryData: ImageQuery | null = null;
```
Comment (guanbinrui) [inline src/app/api/frames/route.tsx:9]:
```suggestion
    const queryData = IMAGE_QUERY_SCHEMA.safeParse(ctx.searchParams);
```

File: src/app/api/frames/vote/route.tsx [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+import { IMAGE_QUERY_SCHEMA, ImageQuery } from '@/constants/zod';
+import { createFrameErrorResponse } from '@/helpers/createFrameErrorResponse';
+import { createFrameSuccessResponse } from '@/helpers/createFrameSuccessResponse';
+import { parseFrameCtxZod } from '@/helpers/parseFrameCtxZod';
+import { parsePollWithZod } from '@/helpers/parsePollWithZod';
+import { getPoll } from '@/services/getPoll';
+import { vote } from '@/services/vote';
+
+export const POST = frames(async (ctx) => {
+    let queryData: ImageQuery | null = null;
```
Comment (guanbinrui) [inline src/app/api/frames/vote/route.tsx:12]:
```suggestion
    const queryData = IMAGE_QUERY_SCHEMA.safeParse(ctx.searchParams)
```

File: src/app/api/frames/route.tsx [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+        const t = createFrameTranslator(locale);
+        const poll = await getPoll(id, source, profileId);
+
+        if (!poll) {
+            throw new Error(t`No poll found via pollId="${id}"`);
+        }
+
+        return createFrameSuccessResponse(poll, queryData);
+    } catch (error) {
+        return createFrameErrorResponse({
```
Comment (guanbinrui) [inline src/app/api/frames/route.tsx:22]: Create a reusable error handler like:

https://github.com/DimensionDev/firefly.mask.social/blob/cd2ddc38f42323ebfa919ea3c218933ac751eef9/src/helpers/withRequestErrorHandler.ts

File: src/app/polls/[id]/page.tsx [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+import { env } from '@/constants/env';
+import { IMAGE_QUERY_SCHEMA } from '@/constants/zod';
+
+interface MetadataProps {
+    params: { id: string };
+    searchParams: { [key: string]: string };
+}
+
+export async function generateMetadata({ params, searchParams }: MetadataProps): Promise<Metadata> {
+    const queryData = IMAGE_QUERY_SCHEMA.parse({
```
Comment (guanbinrui) [inline src/app/polls/[id]/page.tsx:16]: Use safeParse to handle error.

File: src/components/FrameViewer.tsx [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+    const frameData = collectFrameDataFromDocument();
+
+    if (IS_PRODUCTION || !frameData) {
+        return null;
+    }
+
+    return (
+        <div className="flex h-screen w-screen flex-col items-center justify-center">
+            <div className="h-[400px] w-[600px]">
+                <img src={frameData.image} alt="Polls" />
```
Comment (guanbinrui) [inline src/components/FrameViewer.tsx:17]: Use `<Image />`?

File: src/components/PollCard.tsx [hunk 28]
```diff
@@ -0,0 +1,157 @@
...
+                <span
+                    style={{
+                        display: 'flex',
+                        alignItems: 'center',
+                        gap: 8,
+                    }}
+                >
+                    <span style={{ display: 'flex' }}>{choice.name}</span>
+                    {choice.is_select ? (
+                        <img
```
Comment (guanbinrui) [inline src/components/PollCard.tsx:88]: Use `<Image />`?

File: src/config/lensFrame.ts [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+    clientProtocol: {
+        id: 'lens',
+        version: '1.0.0',
+    },
+    handler: {
+        isValidPayload: (body: JSON) => isLensFrameActionPayload(body),
+        getFrameMessage: async (body: JSON) => {
+            if (!isLensFrameActionPayload(body)) {
+                return undefined;
+            }
```
Comment (guanbinrui) [inline src/config/lensFrame.ts:16]:
```suggestion
            if (!isLensFrameActionPayload(body)) return;
```

File: src/components/PollCard.tsx (hunk 28 above)
Comment (dudu0506) [reply src/components/PollCard.tsx:88]: we cant use Image here. This component works in frame.js which doesn't work with Image

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**